consensus/istanbul/core/oracle.go

```diff
@@ -0,0 +1,149 @@
+package core
+
+import (
+	"crypto/rand"
+	"math/big"
+	"reflect"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/consensus/istanbul"
+	"github.com/celo-org/celo-blockchain/consensus/istanbul/algorithm"
+)
+
+type RoundStateOracle struct {
+	rs RoundState
+	c  *core
+
+	roundChangeCertId *algorithm.Value
+	preprepare        *istanbul.Preprepare
+}
+
+func NewRoundStateOracle(rs RoundState, c *core) *RoundStateOracle {
+	return &RoundStateOracle{
+		rs: rs,
+		c:  c,
+	}
+}
+
+func (o *RoundStateOracle) CurrentState() algorithm.State {
+	return algorithm.State(o.rs.State())
+}
+
+func (o *RoundStateOracle) DesiredRound() uint64 {
+	return o.rs.DesiredRound().Uint64()
+}
+
+func (o *RoundStateOracle) checkHeightRoundAndValue(height, round uint64, value algorithm.Value) bool {
+	if height != o.rs.Sequence().Uint64() {
+		return false
+	}
+	if round != o.rs.Round().Uint64() {
+		return false
+	}
+	if o.rs.Proposal() == nil {
+		return false
+	}
+	if o.rs.Proposal().Hash() != common.Hash(value) {
+		return false
+	}
+	return true
+}
+
+func (o *RoundStateOracle) QuorumCommit(height, round uint64, value algorithm.Value) bool {
+	if !o.checkHeightRoundAndValue(height, round, value) {
+		return false
+	}
+	return o.rs.Commits().Size() >= o.rs.ValidatorSet().MinQuorumSize()
+}
+
+func (o *RoundStateOracle) QuorumPrepare(height, round uint64, value algorithm.Value) bool {
+	if !o.checkHeightRoundAndValue(height, round, value) {
+		return false
+	}
+	return o.rs.GetPrepareOrCommitSize() >= o.rs.ValidatorSet().MinQuorumSize()
+}
+
+func (o *RoundStateOracle) QuorumRoundChange() *uint64 {
+	quorumRound := o.c.roundChangeSet.MaxOnOneRound(o.c.current.ValidatorSet().MinQuorumSize())
+	if quorumRound != nil {
+		r := quorumRound.Uint64()
+		return &r
+	}
+	return nil
+}
+
+func (o *RoundStateOracle) FPlus1RoundChange() *uint64 {
+	fRound := o.c.roundChangeSet.MaxRound(o.c.current.ValidatorSet().F() + 1)
+	if fRound != nil {
+		r := fRound.Uint64()
+		return &r
+	}
+	return nil
+}
+
+// SetRoundChangeCertificate maps a round change certificate to an id and
+// returns the id. This allows the algorithm to reference the round change
+// certificate without having a compile time dependency on
+// istanbul.RoundChangeCertificate.
+func (o *RoundStateOracle) SetRoundChangeCertificate(preprepare *istanbul.Preprepare) (*algorithm.Value, error) {
+	// Unset old values
+	o.roundChangeCertId = nil
```

Comment:
It's not really a cache, at most `ValidRoundChangeCert` will be called once for each call to `SetRoundChangeCertificate`.

So the flow would be:

Receive a preprepare -> Set the round change cert -> send the message to the algo -> inspect the round change cert.

All this is happening in a single go-routine, so any time another preprepare is received it will overwrite the last value and the new value will be used just for processing that new preprepare message.